### PR TITLE
Implement callbacks for downloading

### PR DESCRIPTION
**Pull request by Serene-Arc**
Implemented callbacks so now when there's a video downloaded, it'll download in the same place as everything else. Also means that skipping videos will properly work now.

### CODE REVIEW

File: bdfr/resource.py [hunk 1]
```diff
@@ -18,40 +18,52 @@
...
-        except (requests.exceptions.ConnectionError, requests.exceptions.ChunkedEncodingError) as e:
-            logger.warning(f'Error occured downloading from {url}, waiting {current_wait_time} seconds: {e}')
-            time.sleep(current_wait_time)
-            if current_wait_time < max_wait_time:
-                current_wait_time += 60
-                return Resource.retry_download(url, max_wait_time, current_wait_time)
+    def retry_download(url: str) -> Callable:
+        max_wait_time = 300
+
+        def http_download(download_parameters: dict) -> Optional[bytes]:
```
Comment (aliparlakci) [inline bdfr/resource.py:35]: Why not `def http_download(*args, **kwargs)`?

**Comment by aliparlakci** [comment]
I looked at this and it seems reasonable to me. I like FP :)

> it'll download in the same place as everything else. Also means that skipping videos will properly work now.

But can you elaborate more on how this PR improves the areas you mentioned above (and also the ones you didn't mentioned but thought of) just so we are on the same page?

**Comment by Serene-Arc** [comment]
Oops, misclick. Didn't mean to close it.

What's FP?

You know how videos were downloaded at the "find links" stage? This will fix that. It used to be that everything else downloaded in the download stage, but videos had a different flow due to using youtube-dl. So this will make it so that all downloads occur at the same place.

The old way had it downloading before it had a chance to skip any submissions so even if you wanted to skip videos, it would be downloaded anyway. Big waste of bandwidth and time, so this should fix that.

**Review by aliparlakci — APPROVED** [review]
:shipit:

**Pull request merged**